### ISSUE TITLE
Object Detection using Pretrained models

### ISSUE THREAD

**Issue by head-iie-vnr** [Vignana-Jyothi/fdp-embedded-ai-for-ml#5]
Let's make the Raspberry Pi Camera detect objects such as:

👤 Person
🪑 Chair
📱 Cell phone
🖥️ Laptop
🐱 Cat
🐶 Dog
🚗 Car

Instead of drawing arbitrary motion rectangles, the system should produce something like:
```
Person       92%
Cell phone   81%
Laptop       76%
```

**Comment by head-iie-vnr** [comment]
A pretrained model has already been trained by someone else on a large dataset.

### YOLO

YOLO is probably the most popular family for practical real-time object detection.

There are many versions:

YOLOv5
YOLOv8
YOLO11   // very useful and mature choice
YOLO26  // latest 
YOLOX
etc.

### SSD / MobileNet-SSD
Very lightweight models designed particularly for resource-constrained devices.

Accuracy is generally lower than the larger modern YOLO models.

### EfficientDet

Another family designed around the accuracy/speed/compute trade-off.

The EfficientDet family ranges from very small D0 to much larger D7 models.

## Other choices

There are also:

- RT-DETR
- YOLOX
- Faster R-CNN
- Detectron2 models
- MobileNet-based detectors
- TensorFlow Lite detection models


The important lesson is:

**There is no single "best" object detector.**
We choose according to the hardware and application.

**Comment by head-iie-vnr** [comment]
### YOLO11 comes in:

Model | Parameters | COCO mAP | Relative size | Typical idea
-- | -- | -- | -- | --
YOLO11n | 2.6M | 39.5 | 🟢 Tiny | Edge devices
YOLO11s | 9.4M | 47.0 | 🟢 Small | Edge devices
YOLO11m | 20.1M | 51.5 | 🟡 Medium | More powerful systems
YOLO11l | 25.3M | 53.4 | 🟠 Large | High accuracy
YOLO11x | 56.9M | 54.7 | 🔴 Very large | Powerful hardware

The models we're going to use have been trained on COCO, a large object-detection dataset containing 80 object categories.

So our model already knows things like:

👤 person
🚗 car
🚌 bus
🚲 bicycle
🐕 dog
🐈 cat
📱 cell phone
💻 laptop
🪑 chair
🍎 apple
🍌 banana
🎒 backpack
⌚ watch
📺 TV

and many others.

But it doesn't know everything.

So to detect new things like VNRVJIET ID Card **custom training** eventually becomes important.

**Comment by head-iie-vnr** [comment]
### YOLO11n 

YOLO11n has only 2.6 million parameters and 6.5 GFLOPs according to the published model table.

#### Next larger model 
YOLO11s is an excellent candidate too.

It has:

9.4M parameters vs 2.6M for YOLO11n

and published COCO mAP of:

47.0 vs 39.5.

**Comment by head-iie-vnr** [comment]
## High level Architecture 

             Raspberry Pi Camera Module 3
                         │
                         ▼
                  ┌─────────────┐
                  │  Picamera2  │
                  └──────┬──────┘
                         │
                         ▼
                    Video Frame
                         │
                         ▼
                  ┌─────────────┐
                  │   YOLO11n   │
                  │  AI Model   │
                  └──────┬──────┘
                         │
                         ▼
               Object Detection
                         │
             ┌───────────┼───────────┐
             ▼           ▼           ▼
          Person       Phone       Chair
           94%          82%          76%

**Comment by head-iie-vnr** [comment]
mAP@50
"I'll consider the detection correct if the predicted box overlaps the actual box sufficiently."
So mAP50-95 is a fairly strict overall measure of object-detection quality.

COCO contains images with objects from 80 categories, such as:

Therefore:
**YOLO11n = 39.5 mAP**
does NOT mean: "YOLO is correct 39.5% of the time."

That's an important distinction.
It is a benchmark score combining detection precision and bounding-box accuracy over multiple IoU thresholds.

**Comment by head-iie-vnr** [comment]
Do not do
`pip install ultralytics`
For a Raspberry Pi, we should avoid blindly installing the desktop/server CUDA-oriented PyTorch stack.

(py_env1) kp@raspberrypi:~/project1 $ python3 --version
Python 3.13.5
(py_env1) kp@raspberrypi:~/project1 $ python3 -c "import platform; print(platform.machine())"
aarch64
(py_env1) kp@raspberrypi:~/project1 $ pip --version
pip 25.1.1 from /home/kp/project1/py_env1/lib/python3.13/site-packages/pip (python 3.13)


my env is
```
Raspberry Pi 5
        │
        ├── ARM64 / aarch64
        ├── Python 3.13.5
        └── pip 25.1.1
```
My Raspberry Pi OS Trixie repository has native ARM64 packages:
python3-torch        2.6.0
python3-torchvision  0.21.0

**Comment by head-iie-vnr** [comment]
### Installations

sudo apt install -y python3-torch python3-torchvision

> the packages installed by APT should be visible inside py_env1.

(py_env1) kp@raspberrypi:~/project1 $ python3 -c "import torch; print(torch.__version__); print(torch.cuda.is_available())"
2.6.0+debian
False
(py_env1) kp@raspberrypi:~/project1 $ python3 -c "import torchvision; print(torchvision.__version__)"
0.21.0

**Comment by head-iie-vnr** [comment]
```
from picamera2 import Picamera2
from ultralytics import YOLO
import cv2
import time

model = YOLO("yolo11n.pt")

picam2 = Picamera2()

config = picam2.create_preview_configuration(
    main={"size": (640, 480), "format": "RGB888"}
)

picam2.configure(config)
picam2.start()

prev_time = time.time()

while True:
    frame = picam2.capture_array()

    results = model(frame, verbose=False)

    annotated_frame = results[0].plot()

    current_time = time.time()
    fps = 1 / (current_time - prev_time)
    prev_time = current_time

    cv2.putText(
        annotated_frame,
        f"FPS: {fps:.1f}",
        (20, 40),
        cv2.FONT_HERSHEY_SIMPLEX,
        1,
        (0, 255, 0),
        2
    )

    cv2.imshow("YOLO11n - Raspberry Pi 5", annotated_frame)

    if cv2.waitKey(1) & 0xFF == ord("q"):
        break

picam2.stop()
cv2.destroyAllWindows()
```

**Comment by head-iie-vnr** [comment]
<img width="1600" height="1200" alt="Image" src="https://github.com/user-attachments/assets/dfdeef4b-ed92-4baf-85ea-b691b40a3da2" />

Camera Module 3
      ↓
   640 × 480
      ↓
   YOLO11n
      ↓
Object detection
      ↓
┌─────────────────────────┐
│ Keyboard    0.xx        │
│ Laptop      0.xx        │
│ Remote      0.xx        │
│ Bottle      0.xx        │
│ Chair       0.xx        │
└─────────────────────────┘
